docs: Switch from 'breathe' to 'docleaf' for API documentation generation

michaeljones opened the pull request:
This PR switches the Zephyr documentation from using [Breathe](https://github.com/breathe-doc/breathe) to using [Docleaf](https://github.com/docleaf-labs/docleaf) for generating API documentation from Doxygen XML output. 

Docleaf is a re-implementation of Breathe with the aim to improve on Breathe's performance and memory consumption issues. For reference, I am the original creator of Breathe.  Docleaf is largely implemented in Rust with a Python layer to implement the Sphinx extension interface. Builds for Windows, MacOS and Linux and for Python 3.7 to 3.11 are available on [PyPI](https://pypi.org/project/docleaf/). Docleaf is published under the [Parity Public License](https://paritylicense.com/versions/7.0.0).  

I have been working on Docleaf using the Zephyr docs as a test bed and I think the output is at a sufficient standard to replace Breathe. A clean build of the docs on my Ubuntu 22.04 laptop is 2.1x faster than with Breathe and Docleaf performs content checks on the XML files to minimise incremental rebuild times which means that some of the XML copying logic in the doxyrunner code can be removed if you want. I confess I have not done detailed performance analysis of Docleaf to discover further opportunities for improvement. I have not done memory comparisons of the two solutions but assume that the Rust structures holding the XML information are more efficient than the problematic `minidom` module relied upon by Breathe.

Further notes:

- I have tested on Windows 11 and Ubuntu 22.04. I don't have a MacOS device available. 
- When testing on Window 11, either this line (https://github.com/sphinx-doc/sphinx/blob/d3c91f951255c6729a53e38c895ddc0af036b5b9/sphinx/domains/c.py#L3285) or a line like it, passes an Exception straight to `logger.warning` which results in the `warnings_filter.py` code attempting to process an exception object with a regex which then failed so I have adapted the `warnings_filter.py` code to work around that though the might be better approaches. I'm not sure why it wasn't triggered before.
- There are still some warnings printed during the build though less than with Breathe. It might still be sensible to attempt to address them before merging.
- I've used an exact match for the Docleaf version in `requirements-doc.txt` as the project is still young and updates should be taken with some care for the moment. Though no significant breaking changes are planned.
- I've removed `:members:` from all the directive sites as that isn't currently respected by Docleaf which will output all content from the named group which I assume is the desired behaviour. If that is an issue, I can look into adjusting it.  
- I've updated some of the general that referred to Breathe to now refer to Docleaf as it still seems relevant for the moment.
- I have done my best to fix issues in the output but the Zephyr docs are extensive so I may have missed parts.

I hope this all seems reasonable.

## Checklist

- [x] Test on MacOS

michaeljones committed:
doc: Update requirements-doc.txt to use docleaf

Instead of breathe.

I'm not sure of the best version comparison to use. There is no
intention to break the API for docleaf but it is also still young
in some ways and there might be adjustments to the configuration values
that would need to be reflected in the version number as it is still
pre-1.0 that would only be a minor version change.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

michaeljones committed:
doc: Make sure message is a string before regexing it

There are times when the log message is an exception object and that
causes the regex to fail with a TypeError. Here we work around that
by converting the input to a string.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

michaeljones committed:
doc: Replace instances of 'Breathe' with 'Docleaf'

Where valid.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

michaeljones committed:
doc: Remove :members: on doxygengroup directives

As that is not supported by docleaf. We get everything by default.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

michaeljones committed:
doc: Adjust docleaf configuration as needed

We make sure that we have the 'root' and 'xml' defined for the project
which is necessary at the moment for docleaf though the 'root' is only
really used for the 'linkcode' integration.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

michaeljones committed:
doc: Switch to using docleaf for doxygen entities

This switches the Sphinx conf.py file over to using the
'docleaf.doxygen' module and the associated docleaf configuration
entries, replacing the breathe module and config.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

michaeljones commented:
The "Documentation Build (HTML)" test is failing as there is a new warning that isn't covered by the current `known-warnings.txt` patterns. I don't know the preferred way forward here. I can explore getting Docleaf to handle that better (and I assume there might be others if this is failing fast) or we can expand the patterns. I realise the former approach is preferable.

nashif commented:
> I realise the former approach is preferable.

yes, if it can be fixed at the root, that is better than excluding via  file, but to move forward you might want to use the exclusion path I guess.

michaeljones commented:
I will work through them and post progress here. 

The current error contains `enum bt_conn_oob_info::@54 bt_conn_oob_info.type` and which is an nested anonymous enum field which Doxygen doesn't do the best job of representing in its output so it is hard to piece together a clear representation but I should at least try to match the Breathe output (https://docs.zephyrproject.org/latest/connectivity/bluetooth/api/connection_mgmt.html#c.bt_conn_oob_info) or at least not trigger an error.

kartben commented:
This is very cool, thanks for the PR!

> * I have tested on Windows 11 and Ubuntu 22.04. I don't have a MacOS device available.

FWIW it works like a charm on macOS (M2) :)

> * When testing on Window 11, either this line (https://github.com/sphinx-doc/sphinx/blob/d3c91f951255c6729a53e38c895ddc0af036b5b9/sphinx/domains/c.py#L3285) or a line like it, passes an Exception straight to `logger.warning` which results in the `warnings_filter.py` code attempting to process an exception object with a regex which then failed so I have adapted the `warnings_filter.py` code to work around that though the might be better approaches. I'm not sure why it wasn't triggered before.

Ya not sure either but I've had a similar error recently when tinkering with the sphinx_immaterial theme.

michaeljones committed:
doc: Add '--keep-going' flag to sphinx-build

So that we see all the potential errors in CI and not just the first
and can assess how much work there is to do.

This was introduced for debuggin but gmarull has suggested keeping it
in.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

michaeljones committed:
doc: Add more known-warnings

We should probably clear out the old warnings and see what still
applies for docleaf now that we have that instead of breathe.

However for this moment, these are new duplication declarations that
we're seeing errors for and so we'd like to silence them.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

michaeljones commented:
I've marked the PR as 'ready for review' as the tests that ran on the draft seemed to be passing.

michaeljones commented on .github/workflows/doc-build.yml:
```diff
@@ -85,7 +85,7 @@ jobs:
         fi
 
-        DOC_TAG=${DOC_TAG} SPHINXOPTS="-q -W -t publish" make -C doc ${DOC_TARGET}
+        DOC_TAG=${DOC_TAG} SPHINXOPTS="-q -W --keep-going -t publish" make -C doc ${DOC_TARGET}
```
I will remove this tomorrow. I used it for debugging and it doesn't need to stay.

michaeljones replied:
Done

gmarull commented on doc/conf.py:
```diff
@@ -219,6 +219,8 @@
 }
+# Filters out any 'function' or 'variable' members that have 'all caps' names as
+# they are likely unprocessed macro calls
 docleaf_doxygen_skip = ["members:all_caps"]
```
fixup original commit

gmarull commented on scripts/requirements-doc.txt:
```diff
@@ -1,6 +1,6 @@
 # DOC: used to generate docs
 
-docleaf==0.6.0
+docleaf==0.7.0
```
ditto

gmarull commented on .github/workflows/doc-build.yml:
```diff
@@ -85,7 +85,7 @@ jobs:
         fi
 
-        DOC_TAG=${DOC_TAG} SPHINXOPTS="-q -W -t publish" make -C doc ${DOC_TARGET}
+        DOC_TAG=${DOC_TAG} SPHINXOPTS="-q -W --keep-going -t publish" make -C doc ${DOC_TARGET}
```
drop? I wonder if we should actually default to this, I was not aware of this option!

gmarull commented on scripts/requirements-doc.txt:
```diff
@@ -1,6 +1,6 @@
 # DOC: used to generate docs
 
-docleaf==0.7.0
+docleaf==0.8.0
```
ditto

gmarull reviewed (changes requested):
note: try to preserve bisectability with the changes

michaeljones commented on doc/known-warnings.txt:
```diff
@@ -16,3 +16,6 @@
 .*Duplicate C declaration.*\n.*'\.\. c:.*:: .*struct in6_addr.*'.*
 .*Duplicate C declaration.*\n.*'\.\. c:.*:: .*struct net_if.*'.*
 .*Duplicate C declaration.*\n.*'\.\. c:struct:: bt_ots_init'.*
+.*Duplicate C declaration.*\n.*'\.\. c:member:: enum *bt_mesh_dfd_upload_phase bt_mesh_dfd_srv.phase'.*
+.*Duplicate C declaration.*\n.*'\.\. c:member:: struct *bt_mesh_blob_xfer bt_mesh_dfu_cli.blob'.*
+.*Duplicate C declaration.*\n.*'\.\. c:member:: struct *net_if *\* net_if_mcast_monitor.iface'.
```
Initial impression is that the root of the issue at times seems to be that the Sphinx C-domain does not have good handling for when structs, functions and other entities have the same name. This is allowed in C and present a few times in the Zephyr code base. This leads to duplicate targets being generated and triggering warnings from the Sphinx C Domain  code.

It is mentioned here: https://github.com/sphinx-doc/sphinx/issues/8241#issuecomment-701245203 in an issue which goes on to reference this pull request: https://github.com/sphinx-doc/sphinx/pull/8313 as the solution but that seems to have been open for a long time. 

As the references are parsed and stored by the C Domain which then generates the warnings for clashes, I don't see there being a clear way to work around them other than the current regexes. The question then becomes the difference between Breathe and Docleaf and that comes down to how much of the Doxygen XML the two systems attempt to render. In various edge cases, the XML becomes harder to make clear sense of and so the two systems potentially draw the lines at different points in different places. Neither are perfect. 

I will work through the various cases (as highlighted by deleting all the lines in known-warnings.txt file) and document why each appears to be happening and attempt to improve Docleaf where possible. Any due to name clashes are potentially going to have to stay but I can see at least one case where field name is clashing with a field in a nested structure and that might be possible to avoid.

michaeljones replied:
I've added a new commit which adds categories to the known-warnings.txt file attempting to indicate the source of the various issues. Some are due to entities sharing the same name and the Sphinx issues indicated before. Some are due to fields with the same name appearing in multiple anonymous structs. Some are due to Doxygen not indicating `extern` declarations in the XML output and so Docleaf and Breathe are unable to filter them out in any fashion. I have raised a ticket for this: https://github.com/doxygen/doxygen/issues/10163

It is possible to use the Doxygen `@cond` syntax to have it ignore parts of the file and so we could use that to explicit stop it processing the `extern` declarations but that would introduce more changes than the current known-warnings.txt patterns.

Finally the new lines are due to Docleaf attempting to render fields from anonymous structs which then clash with fields in the parent struct. I believe this didn't happen in Breathe as it doesn't attempt to render the members in some of these situations. For example, the breathe output for `bt_mesh_dfd_srv` is:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/5390/126e6c61-9809-49b5-908a-ba9accc7ff76)

Whilst the docleaf output is:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/5390/005e2d6e-ba31-4552-9c42-ff87a05e11f8)

With the actual code being: https://github.com/zephyrproject-rtos/zephyr/blob/db1a718341a3724b82c052e6e6b7db19251f7a22/include/zephyr/bluetooth/mesh/dfd_srv.h#L117-L136

Which shows that the anonymous struct members incorrectly appear in the body of the main struct. Unfortunately this is because the Doxygen xml output ([structbt__mesh__dfd__srv.zip](https://github.com/zephyrproject-rtos/zephyr/files/11930141/structbt__mesh__dfd__srv.zip)) doesn't distinguish between the two levels. It flattens everything. I'm not sure what is better or worse. Leaving out the details or having them poorly represented. I haven't yet looked to see if there is a doxygen issue for this but I fear it might involve changes to their XML structures to better support this and that seems like it is possibly unlikely.

michaeljones replied:
I would welcome input as to what we might want to address or account for at this stage or if the current state is acceptable in some fashion for moving forwards.

michaeljones committed:
doc: Log unused log filter patterns

This allows us to identify any patterns that we can remove or that
we didn't realise are no longer in use. This might happen if issues
within doxygen or docleaf are resolved.

This allows us to remove the pattern:

  .*Duplicate C declaration.*\n.*'\.\. c:.*:: uint16_t id'.*

which does not match anything in the current set up.

We also split the filter patterns in known-warnings out into different
sections depending on their cause.

Also extend the pattern parser to ignore empty lines so that we can
have some formatting in the known-warnings file.

Signed-off-by: Michael Jones <m.pricejones@gmail.com>

gmarull reviewed (approved):
lgtm, thanks!
